src/main/scala/mesosphere/marathon/api/HttpTransferMetricsHandler.scala

```diff
@@ -0,0 +1,130 @@
+import javax.ws.rs.core.HttpHeaders
+import mesosphere.marathon.metrics.{Counter, Metrics, ServiceMetric}
+import org.eclipse.jetty.server.Request
+import org.eclipse.jetty.server.handler.AbstractHandler
+import org.eclipse.jetty.server.{HttpChannelState, _}
+
+import HttpTransferMetricsHandler._
+
+/* Container for HTTP Metrics
+ */
+trait HttpTransferMetrics {
+  /**
+    * Metric representing the total HTTP entity gzipped bytes read (does not include HTTP headers)
+    */
+  val gzippedBytesReadMetric: Counter
+
+  /**
+    * Metric representing the total HTTP entity gzipped bytes written (does not include HTTP headers)
+    */
+  val gzippedBytesWrittenMetric: Counter
+
+  /**
+    * Metric representing the total HTTP entity bytes read (does not include HTTP headers)
+    */
+  val bytesReadMetric: Counter
+
+  /**
+    * Metric representing the total HTTP entity bytes written (does not include HTTP headers)
+    */
+  val bytesWrittenMetric: Counter
+}
+
+/**
+  * Implementation class for HttpMetrics
+  *
+  * This class is structured and named this way for backwards compatibility, due to the fact that our metrics are
+  * currently (and lamentably) tied to classes.
+  */
+class HTTPMetricsFilter extends HttpTransferMetrics {
+  val gzippedBytesReadMetric = Metrics.counter(ServiceMetric, getClass, "gzippedBytesRead")
+  val gzippedBytesWrittenMetric = Metrics.counter(ServiceMetric, getClass, "gzippedBytesWritten")
+  val bytesReadMetric = Metrics.counter(ServiceMetric, getClass, "bytesRead")
+  val bytesWrittenMetric = Metrics.counter(ServiceMetric, getClass, "bytesWritten")
+}
+
+class HttpTransferMetricsHandler(httpMetrics: HttpTransferMetrics) extends AbstractHandler with StrictLogging {
+  val metricsListener: AsyncListener = new AsyncListener() {
+    def onError(event: javax.servlet.AsyncEvent): Unit = {}
+
+    def onTimeout(event: javax.servlet.AsyncEvent): Unit = {}
+
+    def onStartAsync(event: javax.servlet.AsyncEvent): Unit = {
+      event.getAsyncContext().addListener(this);
+    }
+
+    override def onComplete(event: AsyncEvent): Unit = event match {
+      case ace: AsyncContextEvent =>
+        updateResponse(ace.getHttpChannelState.getBaseRequest)
+      case _ =>
+        logger.error(s"${event.getClass} wasn't an AsyncContextEvent (We should see this)")
+    }
+  }
+
+  override def handle(path: String, baseRequest: Request, request: HttpServletRequest, response: HttpServletResponse): Unit = {
+    val state = baseRequest.getHttpChannelState
+
+    if (state.isInitial()) {
+      if (state.isSuspended())
+        state.addListener(metricsListener)
+      else
+        updateResponse(baseRequest)
+    }
+  }
+
+  private def isGzip(contentEncodingHeader: String): Boolean =
+    (contentEncodingHeader != null) && contentEncodingHeader.equalsIgnoreCase("gzip")
+
+  protected def updateResponse(request: Request) =
+    if (!isExcluded(request)) {
+      val bytesRead = request.getContentLengthLong
```

Comment:
I guess we counted only request/response body sizes before as well?.. And it seems that it is non-trivial to count true request/response sizes.

Reply:
https://github.com/mesosphere/marathon/pull/6288